### ISSUE TITLE
UsePagination / UseFiltering causes dropped data for 1:N relationships

### ISSUE THREAD

**Issue by zaneclaes** [ChilliCream/hotchocolate#5131]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I am using pagination/filtering/sorting with an `IEnumerable` output from my MySQL database (using Pomelo).

When I issue a GraphQL query to filter upon some query input, the filter is matched and the correct items are returned, but the nested arrays are cleared (empty).

### Steps to reproduce

GraphQL endpoint:
```
[UseFurDataContext] [UsePaging] [UseFiltering] [UseSorting]
  public IEnumerable<Furball> SearchFurballs([Service] FurContext context) => context.Data.Furballs.IncludeDefaultFurball();
```

Where `Data` is a `DbContext`.  The `IncludeDefaultFurball` performs `.Include(fb => fb.Inventory).ThenInclude(c => c.Items)`. Thus, I am able to successfully filter the output with the query:

```
query {
  searchFurballs(where: { inventory: { items: { some: { itemId: { eq: 66561 } } } } }) {
```

The top level items returned are correct, but the nested `inventory.items` are missing from the output:

```
{
  "data": {
    "searchFurballs": {
      "nodes": [
        {
          "id": "0x010201070300070a07081800",
          "inventory": {
            "id": "0x010201070300070a07081800",
            "items": []
          }
        },
        {
          "id": "0x01040713040001060a010400",
          "inventory": {
            "id": "0x01040713040001060a010400",
            "items": []
          }
        },
```

I have verified that there should be `items` present by also double checking the database, as well as performing other queries on the same objects. 

The same bug also happens on any other nested `List` style field that I attempt to filter upon.

Curiously, **the bug can be worked around** by adding a ` id: { neq: "" }, ` to the `where` clause. Suddenly, the correct data is returned when I run this query: 

```
query {
  searchFurballs(where: { id: { neq: "" }, inventory: { items: { some: { itemId: { eq: 66561 } } } } }) {
```

### Product

Hot Chocolate

### Version

12.9.0

**Comment by PascalSenn** [comment]
@zaneclaes  i guess you need to use projections to project related entites

**Comment by PascalSenn** [comment]
this is the way we do includes

**Comment by zaneclaes** [comment]
@PascalSenn I've never been able to use Projections in any of my HotChocolate projects due to your comment here: https://github.com/ChilliCream/hotchocolate/issues/1658#issuecomment-716068068

I notice now that it was expected to be fixed last year with https://github.com/ChilliCream/hotchocolate/pull/3650 , but I'm still getting the same error with version `12.9.0`:

C# code:
```
[UseFurDataContext] [UsePaging] [UseProjection] [UseFiltering] [UseSorting]
  public IQueryable<Furball> SearchFurballs([Service] FurContext context) => context.Data.Furballs;
```

GraphQL result:
```
{
  "errors": [
    {
      "message": "Can't compile a NewExpression with a constructor declared on an abstract class",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "searchFurballs"
      ],
      "extensions": {
        "code": "NotFound"
      }
    }
  ],
  "data": {
    "searchFurballs": null
  }
}
```

**Comment by PascalSenn** [comment]
the issue is resolved in the way that you can now project abstract types. the limitation you run into is still there and  is not really fixable with how projections work at the moment. I suppose you have a abstract Furball somewhere in your abstractions and a concrete implementation somwhere else, this makes it no possible to project like `new Furball() { id = x.id}` as the class is abstract. We also do not know the concrete types so we can also not project into `new ConcreteFurball() { Id =x.Id}`. 

in v13 we provide a context for Projections. so that you have easier access to the selected fields (and can include manually)

**Comment by PascalSenn** [comment]
what's weird is indeed that the includes are ignored. How does the expression look like? 

You can flilter in the resolver to have a look at the expression:
```csharp
[UseFurDataContext] [UsePaging] [UseFiltering] [UseSorting]
  public IEnumerable<Furball> SearchFurballs([Service] FurContext context, IResolverContext rc) => context.Data.Furballs.IncludeDefaultFurball().Filter(rc).Sort(rc); <-- what is this expression
```

**Comment by zaneclaes** [comment]
>  I suppose you have a abstract Furball somewhere in your abstractions and a concrete implementation somwhere else

No, I do not. The base `abstract` class does not share a name with the concrete implementation.

> what's weird is indeed that the includes are ignored. 

I don't think that's true. **Notice** that the `where` clause is **filtering on the included fields**!! There is no way the query would have returned any data at all if the Includes were ignored. HotChocolate **must** be clearing the included data at some point between the `where` clause and returning the data.

> You can flilter in the resolver to have a look at the expression:

Where should I see the expression? I added that but don't see any logs or anything.

**Comment by zaneclaes** [comment]
It seems the problem with Projections is _nested_ abstract relationships.

In the query I gave, the `items` are an array of `abstract GameItem` objects. If I remove `inventory.items` from the GraphQL selection set, the query works:
<img width="1164" alt="Screen Shot 2022-06-07 at 7 24 05 AM" src="https://user-images.githubusercontent.com/2084821/172391278-8d1eb30f-c2ed-4c20-97c7-0da1041421db.png">

And including the `items` breaks it:
<img width="1676" alt="Screen Shot 2022-06-07 at 7 24 16 AM" src="https://user-images.githubusercontent.com/2084821/172391329-86433f0a-94df-4049-8b17-af68a737d17d.png">

Where items are:
`public List<GameItem> Items { get; set; } = new List<GameItem>();`

And `abstract GameItem` is a unique class name (not shared by any other class).

**Comment by PascalSenn** [comment]
@zaneclaes  is GameItem a graphql interface/union?

**Comment by PascalSenn** [comment]
hm...  i see you problem. 
So projecting gamestats would resolve the whole list of gamestats, so even if that would work, then we would load unnecessary columns. 
In v13 we add support for expression in filtering and sorting. We also want to add expression support to projection, so that you could actually express your resolver as a expression (`GameStats.Last`). This would make you able to project these things.
The Problem we have there is that currently we project directly into the model, essentially `.Select(x => new Furball { Id = x.Id)`.  This works great for Models, but as soon as you have fancy resolvers on them, we have to backing field to project. 
What we plan to do is to rewrite how projections work internally so that we can project everything we want. 
But to implement this we need a bunch of changes to the execution engine.

**Comment by michaelstaib** [comment]
Hey,

I have tested this on main.

Current main commit: `f64f691347`.

What I tested:

```csharp
public sealed class Query
{
    [UsePaging]
    [UseFiltering]
    [UseSorting]
    public IQueryable<Parent> GetParents(AppDbContext db) => db.Parents;
}

public sealed class Parent
{
    public int Id { get; set; }
    public ICollection<Child> Children { get; set; } = new List<Child>();
}

[Fact]
public async Task Paging_Filtering_Sorting_Does_Not_Clear_Selected_Child_Collection()
{
    var result = await executor.ExecuteAsync(
        """
        {
          parents(where: { children: { some: { name: { eq: \"a\" } } } }) {
            nodes {
              id
              children { id name }
            }
          }
        }
        """);
}
```

Result:
- Nested selected collection data stayed intact after filtering.
- I could not reproduce the clearing/emptying behavior on current main.
